modules-js/react-fleet/src/form-elements/inputs/MemorableDateInput.tsx

```diff
@@ -0,0 +1,458 @@
+import React from 'react';
+
+import { css } from 'emotion';
+
+interface Props {
+  initialDate?: string;
+  earliestDate?: string;
+  latestDate?: string;
+  onlyAllowPast?: boolean;
+  onlyAllowFuture?: boolean;
+  legend: React.ReactChild;
+  handleDate: (newDate: Date | null) => void;
+}
+
+interface State {
+  hasFocus: boolean;
+  error: string;
+  date: Date | null;
+  fields: {
+    month: number;
+    day: number;
+    year: number;
+  };
+}
+
+/**
+ * Component to collect a memorable date (such as a birth date) from a user.
+ *
+ * An element must be passed in as the legend, commonly a header <h1-h6>.
+ *
+ * Default behavior is to accept any date. A text string representing a date
+ * can be passed into earliestDate and/or latestDate to limit acceptable input.
+ * Format: MM-DD-YYYY, MM/DD/YYYY, etc. (https://tools.ietf.org/html/rfc2822#section-3.3)
+ *
+ * An initial date can be passed in via the initialDate property.
+ *
+ * onlyAllowPast: If true, will only accept yesterday or earlier for a date.
+ * onlyAllowFuture: If true, will only accept tomorrow or later for a date.
+ *
+ * Whenever the date is updated, this.props.handleDate is called.
+ *
+ * *Please note that all dates are relative to the user’s computer.
+ *
+ * Inspired by https://design-system.service.gov.uk/components/date-input/
+ */
+export default class MemorableDateInput extends React.Component<Props, State> {
+  readonly earliest: Date | null;
+  readonly latest: Date | null;
+  readonly initial: Date | null;
+
+  private timeout: any;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.initial = props.initialDate ? new Date(props.initialDate) : null;
+
+    checkIfPropsValid(props);
+
+    this.earliest = setDateLimit(
+      props.earliestDate || '',
+      !!props.onlyAllowFuture
+    );
+    this.latest = setDateLimit(props.latestDate || '', !!props.onlyAllowPast);
+
+    this.state = {
+      hasFocus: false,
+      error: '',
+      date: this.initial,
+      fields: this.initial
+        ? {
+            month: this.initial.getMonth(),
+            day: this.initial.getDate(),
+            year: this.initial.getFullYear(),
+          }
+        : ({
+            month: '',
```

Comment:
You should settle on an accurate type of the `fields`. I'd recommend `number | null` but you could do `string` instead.

Reply:
If `null` is used, React throws
```console
Warning: `value` prop on `input` should not be null. Consider using an empty string to clear the component or `undefined` for uncontrolled components.
```
`undefined` causes it to treat the fields like uncontrolled components and it gets crankier. 😜

But for sure, I’ll add the overlooked `Fields` type!

Reply:
In the logic of your component, the fields should probably be `number`s. You’ll need to translate that into `string`s when rendering and back to `number`s when the `value` changes, since `<input>`’s `value` attribute is a `string`.

Or, just keep them in your `State` as `string`s and translate when you’re constructing the `Date`.

Whichever you choose isn’t as important as being clear about the types.